process/dnsProcessor.go

```diff
@@ -0,0 +1,124 @@
+package process
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/hashing/keccak"
+	"github.com/ElrondNetwork/elrond-proxy-go/process/helpers"
+)
+
+var usernameHasher = &keccak.Keccak{}
+
+const (
+	lowUsernameLengthBoundary  = 3
+	highUsernameLengthBoundary = 20
+	usernameSuffix             = ".elrond"
+)
+
+// DnsProcessor handles dns operations
+type DnsProcessor struct {
+	sortedEncodedAddresses []string
+	pubKeyConverter        core.PubkeyConverter
+}
+
+// NewDnsProcessor initializes all the inner components and returns the newly created instance of DnsProcessor
+func NewDnsProcessor(pubKeyConverter core.PubkeyConverter) (*DnsProcessor, error) {
+	if check.IfNil(pubKeyConverter) {
+		return nil, ErrNilPubKeyConverter
+	}
+
+	dp := &DnsProcessor{
+		pubKeyConverter: pubKeyConverter,
+	}
+
+	err := dp.computeDnsAddresses()
+	if err != nil {
+		return nil, err
+	}
+
+	return dp, nil
+}
+
+// GetDnsAddresses returns all the dns addresses
+func (dp *DnsProcessor) GetDnsAddresses() ([]string, error) {
+	return dp.sortedEncodedAddresses, nil
+}
+
+// GetDnsAddressForUsername returns the corresponding dns address for the provided username
+func (dp *DnsProcessor) GetDnsAddressForUsername(username string) (string, error) {
+	err := checkUsername(username)
+	if err != nil {
+		return "", err
+	}
+
+	if !strings.HasSuffix(username, usernameSuffix) {
+		username += usernameSuffix
+		if len(username) > highUsernameLengthBoundary {
+			return "", ErrInvalidUsernameLength
+		}
+	}
+
+	usernameBytes := usernameHasher.Compute(username)
+	if len(usernameBytes) < 32 {
+		return "", ErrHashedUsernameBelowLimit
+	}
+
+	address := dp.sortedEncodedAddresses[usernameBytes[31]]
+
+	return address, nil
+
+}
+
+func checkUsername(username string) error {
+	usernameLength := len(username)
+	if !(usernameLength > lowUsernameLengthBoundary && usernameLength < highUsernameLengthBoundary) {
+		return ErrInvalidUsernameLength
+	}
+
+	for _, c := range username {
+		if (c < 'a' || c > 'z') && c != '.' {
```

Comment:
Also numbers should be allowed. 
Could we somehow extract this validation from elrond and not have it 2 ways in 2 projects?

Reply:
replaced the conditions and used a regex instead of that for->if and added more tests. @raduchis from my knowledge, there isn't a common function to compute the validity. I think the actual checks are done in the dns sc which is written in rust.

Reply:
yes, you are right, it was a test that checked the validity of the username